ユーザー管理機能の実装

swatanabe1122 opened the pull request:
# What
新規登録/ログイン/ログアウトができる。
# Why
ユーザー管理機能を実装するため。
### 以下を、1回目のレビュー後に最新の画像や動画に変更してます。
 必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画 
https://gyazo.com/9deea5bd3a0ae09f75b31c1a51b0e632
 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/e5f722dc64becdf3fd6650ae9783bce2
 必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/2f08dce19c5afddabd234105d8252a3c
 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/15bfb9aeaf654fdbffd93f4e87f0b5ca
 ログアウトに成功する動画
https://gyazo.com/6a881dff28e6ea21b4f8f240c0127043
 テストに成功する画像
https://gyazo.com/e5e3658b7782d528cd0c2f9b6170df7f

swatanabe1122 committed:
ユーザー管理機能の実装

deviseのインストール、テーブルを作成

swatanabe1122 committed:
Revert "ユーザー管理機能の実装"

This reverts commit 82ce2553db5780847cd44165ba17a451e214bbd5.

swatanabe1122 committed:
Revert "Revert "ユーザー管理機能の実装""

This reverts commit fa8f9f3d435420a8cf0d4e32b62219fb0f280abd.

swatanabe1122 committed:
Revert "ユーザー管理機能の実装"

This reverts commit 82ce2553db5780847cd44165ba17a451e214bbd5.

swatanabe1122 committed:
Revert "Revert "ユーザー管理機能の実装""

This reverts commit 3fdf834a4a6159871e7b20fb499ab6db42bf300d.

swatanabe1122 committed:
バリデーションの設定

swatanabe1122 committed:
新規登録の画面表示作成

オフラインで作業できなかったので、認証も修正

swatanabe1122 committed:
ユーザー認証機能の実装

ログインしていない場合はログイン画面に遷移する記述

swatanabe1122 committed:
ユーザー管理機能

ユーザーがログインしていなければ、そのユーザーをログイン画面に遷移させる

swatanabe1122 committed:
ユーザー管理機能

ログインの有無で表示が変わるように分岐、　　　　　　　　　　　　　　　　　　　　　CSSファイル追加してなかったので追加

swatanabe1122 committed:
ログアウト機能の実装

swatanabe1122 committed:
ユーザー管理機能の実装

バリデーションの設定

swatanabe1122 committed:
ユーザー管理機能

テストコードの記述

swatanabe1122 committed:
ユーザー管理機能のテストコード

swatanabe1122 committed:
ユーザ管理機能

エラー発生時にメッセージが表示されないので、修正。

swatanabe1122 committed:
ユーザー管理機能

Rubocop使ってなかったので使用

maimai0401 commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,21 @@
 class ApplicationController < ActionController::Base
+  before_action :authenticate_user!
```
◎authenticate_user!メソッドを削除しましょう。

■理由
ユーザー管理機能の実装要件において未ログイン時のページ遷移制限は求められていないためです。

maimai0401 commented on app/models/user.rb:
```diff
@@ -0,0 +1,23 @@
+
+  # パスワードのバリデーション
+  # 半角英数字混合
+  validates :password, format: { with: /\A(?=.*?[a-zA-Z])(?=.*?\d)[a-zA-Z\d]+\z/ }, on: :create
```
```
, on: :create
```
上記記述は必要でしょうか？
今一度検証を行いましょう。

maimai0401 commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -1,7 +1,7 @@
 <%= render "shared/second-header"%>
 
 <%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

maimai0401 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,133 @@
+        expect(@user.errors.full_messages).to include('Password is too short (minimum is 6 characters)')
+      end
+
+      it 'パスワードは半角英数字混合であること' do
```
◎異常系テストを追加しましょう。

■理由
下記要件を満たして頂くためです。
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない

※パスワードにつきましては上記それぞれの項目に分けてexampleの生成が必要です。

maimai0401 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,133 @@
+        @user.last_name_kana = 'たろう'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('First name kana is invalid', 'Last name kana is invalid')
+      end
```
```
    it 'お名前(全角)が空では登録できない' do
        @user.first_name = ''
        @user.last_name = ''
        @user.valid?
        expect(@user.errors.full_messages).to include("First name can't be blank", "Last name can't be blank")
      end

      it 'お名前(全角)は全角（漢字・ひらがな・カタカナ）であること' do
        @user.first_name = 'abc'
        @user.last_name = '123'
        @user.valid?
        expect(@user.errors.full_messages).to include('First name is invalid', 'Last name is invalid')
      end

      it 'お名前カナ(全角)が空では登録できない' do
        @user.first_name_kana = ''
        @user.last_name_kana = ''
        @user.valid?
        expect(@user.errors.full_messages).to include("First name kana can't be blank", "Last name kana can't be blank")
      end

      it 'お名前カナ(全角)は全角カタカナであること' do
        @user.first_name_kana = 'やまだ'
        @user.last_name_kana = 'たろう'
        @user.valid?
        expect(@user.errors.full_messages).to include('First name kana is invalid', 'Last name kana is invalid')
      end
```
1つのit内では１つのカラムに関するテストのみ行いましょう。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください。
どうぞよろしくおねがいします。

swatanabe1122 committed:
プルリクエスト後に修正。

authenticate_user!メソッドを削除。
バリデーションが独立してテストされるように、ブロック内の記述を、名字と名前に独立。
異常系テストを追加。
不要な記述の削除

swatanabe1122 committed:
1回目のレビュー後に、Rubocopで修正後

swatanabe1122 commented:
### 1回目のレビュー後の最新の画像と動画。
 必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画 
https://gyazo.com/9deea5bd3a0ae09f75b31c1a51b0e632
 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/e5f722dc64becdf3fd6650ae9783bce2
 必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/2f08dce19c5afddabd234105d8252a3c
 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/15bfb9aeaf654fdbffd93f4e87f0b5ca
 ログアウトに成功する動画
https://gyazo.com/6a881dff28e6ea21b4f8f240c0127043
 テストに成功する画像
https://gyazo.com/e5e3658b7782d528cd0c2f9b6170df7f

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。

こちら`LGTM`とさせていただきます！実装お疲れさまでした。

【注意】
```ruby
①（次のブランチを切る前に）プルリクエストのMergeとmainブランチへのPullを忘れないようにしましょう。
②次のフェーズに進む際は必ず新しいブランチを作ってから作業しましょう。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
引き続き頑張りましょう！応援しております！
